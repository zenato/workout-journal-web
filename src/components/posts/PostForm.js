import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray, reduxForm } from 'redux-form'
import classNames from 'classnames/bind';
import { Button, Input } from 'components/form';
import Performance from 'components/posts/Performance';
import styles from './PostForm.scss';

const cx = classNames.bind(styles);

const validate = ({ workoutDate, performances }) => {
  const errors = {};

  if (!workoutDate) {
    errors.workoutDate = 'Required';
  }

  const performanceArrayErrors = [];
  (performances || []).forEach((performance, idx) => {
    const performanceErrors = {};
    if (performance) {
      if (!performance.event) {
        performanceErrors.event = 'Required';
      }
      if (!performance.value) {
        performanceErrors.value = 'Required';
      } else if (!/^[0-9]{1,3}$/.test(performance.value)) {
        performanceErrors.value = 'Must be numeric, 0-999';
      }
      ['set1', 'set2', 'set3', 'set4', 'set5'].forEach(fieldName => {
        const value = performance[fieldName];
        if (value && !/^[0-9]{1,3}$/.test(value)) {
          performanceErrors[fieldName] = 'Must be numeric, 0-999';
        }
      });
    }
    if (Object.keys(performanceErrors).length) {
      performanceArrayErrors[idx] = performanceErrors;
    }
  });

  if (performanceArrayErrors.length > 0) {
    errors.performances = performanceArrayErrors;
  }

  return errors;
};

const renderPerformances = ({ fields, meta: { error, submitFailed }, events, formValues }) => (
  <div className={cx('item')}>
    <label>Perf.</label>
    <div className={cx('field')}>
      <Button onClick={() => fields.push({})} value="Add" />
      <table>
        <colgroup>
          <col width="*" />
          <col width="8%" />
          <col width="5%" />
          <col width="8%" />
          <col width="8%" />
          <col width="8%" />
          <col width="8%" />
          <col width="8%" />
          <col width="8%" />
          <col width="5%" />
        </colgroup>
        <thead>
        <tr>
          <th>Name</th>
          <th colSpan={2}>Val.</th>
          <th>1</th>
          <th>2</th>
          <th>3</th>
          <th>4</th>
          <th>5</th>
          <th>Tot. / Vol.</th>
          <th></th>
        </tr>
        </thead>
        <tbody>
        {fields.map((name, index) => (
          <Performance
            key={index}
            name={name}
            events={events}
            values={formValues.performances[index] || {}}
            onDelete={() => fields.remove(index)}
          />
        ))}
        </tbody>
      </table>
    </div>
  </div>
);

class PostForm extends Component {
  handleDelete = (e) => {
    e.preventDefault();
    this.props.onDelete();
  };

  handleMoveList = (e) => {
    e.preventDefault();
    this.props.onMoveList();
  };

  render() {
    const { initialValues, handleSubmit, loading, error, events, formValues } = this.props;
    return (
      <form onSubmit={handleSubmit} className={cx('post-form')}>
        <Field type="date" name="workoutDate" label="Workout Date" component={Input} />
        <FieldArray name="performances" events={events} formValues={formValues} component={renderPerformances} />
        <Field type="text" name="remark" label="Remark" component={Input} />

        {!loading && error && (
          <div className={cx('error')}>
            <span>Oops, An expected error seems to have occurred.</span>
          </div>
        )}

        <div className={cx('tool')}>
          <Button type="submit" value="Save" className={cx('primary')} />
          <Button value="List" onClick={this.handleMoveList} />
          {initialValues.id && (
            <Button value="Delete" onClick={this.handleDelete} />
          )}
        </div>
      </form>
    );
  }
}

PostForm.propTypes = {
  initialValues: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  formValues: PropTypes.object.isRequired,
  events: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.object,
  onSubmit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onMoveList: PropTypes.func.isRequired,
};

export default reduxForm({
  form: 'postForm',
  validate,
})(PostForm);

