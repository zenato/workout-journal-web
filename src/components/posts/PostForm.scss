@import 'common';

.post-form {
  padding: 10px;

  .item {
    display: flex;
    margin-bottom: 5px;
    align-items: center;

    label {
      flex: 2;
      display: inline-block;
      vertical-align: middle;
      text-align: right;
      font-weight: bold;
      margin-right: 5px;
    }

    .field {
      flex: 10;
      padding-right: 10px;
      vertical-align: middle;
      overflow-x: auto;
    }

    .error-message {
      color: #f00;
      font-size: 0.7rem;
      padding: 5px;
    }
  }

  .tool {
    text-align: center;

    * {
      margin-left: 3px;
    }
  }

  .error {
    color: #f00;
    text-align: center;
    font-size: 0.7rem;
    padding: 5px;
  }
}
