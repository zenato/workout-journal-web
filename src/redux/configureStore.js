import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import penderMiddleware from 'redux-pender';
import modules from './modules';

const isDevelopment = process.env.NODE_ENV === 'development';

const composeEnhancers = isDevelopment
  ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  : compose;

const configureStore = initialState => {
  const store = createStore(
    modules,
    initialState,
    composeEnhancers(applyMiddleware(thunk, penderMiddleware())),
  );

  if (module.hot) {
    module.hot.accept('./modules', () => {
      const nextRootReducer = require('./modules').default;
      store.replaceReducer(nextRootReducer);
    });
  }

  return store;
};

export default configureStore;
