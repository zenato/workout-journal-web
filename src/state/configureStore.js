import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import thunk from 'redux-thunk'
import penderMiddleware from 'redux-pender'
import reducers from './reducers'
import rootSaga from './sagas'

const isDevelopment = process.env.NODE_ENV === 'development'

const composeEnhancers = isDevelopment
  ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose
  : compose

const configureStore = initialState => {
  const sagaMiddleware = createSagaMiddleware()
  const store = createStore(
    reducers,
    initialState,
    composeEnhancers(applyMiddleware(sagaMiddleware, thunk, penderMiddleware())),
  )

  if (module.hot) {
    module.hot.accept('./reducers', () => {
      const nextRootReducer = require('./reducers').default
      store.replaceReducer(nextRootReducer)
    })
  }

  sagaMiddleware.run(rootSaga);
  return store
}

export default configureStore
