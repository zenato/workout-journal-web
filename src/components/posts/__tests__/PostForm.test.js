import React from 'react';
import { Provider } from 'react-redux';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux/configureStore';
import PostForm from '../PostForm';

configure({ adapter: new Adapter()});

const store = configureStore();

const Test = (props) => (
  <Provider store={store} {...props} />
);

const events = [
  {
    id: '1',
    name: 'test',
    value: 10,
    unit: 'g',
    remark: 'test',
  },
];

const initialValues = {
  id: '1',
  workoutDate: new Date().getTime(),
  remark: 'test',
  performances: [
    {
      event: events[0],
      value: 20,
      set1: 1,
      set2: 2,
      set3: 3,
      set4: 4,
      set5: 5,
    },
  ],
};

it('renders without errors', () => {
  shallow(
    <Test>
      <PostForm
        events={events}
        initialValues={initialValues}
        error={null}
        loading={false}
        onSubmit={() => {}}
        onDelete={() => {}}
        onMoveList={() => {}}
      />
    </Test>
  );
});

it('simulate submit after change form', () => {
  const onSubmit = jest.fn();
  const component = mount(
    <Test>
      <PostForm
        events={events}
        initialValues={initialValues}
        formValues={initialValues}
        error={null}
        loading={false}
        onSubmit={onSubmit}
        onDelete={() => {}}
        onMoveList={() => {}}
      />
    </Test>
  );

  // Changed input form
  component.find('input[name="workoutDate"]').simulate('change', {
    target: {
      name: 'workoutDate',
      value: '2017-08-08',
    },
  });
  component.find('input[name="remark"]').simulate('change', {
    target: {
      name: 'remark',
      value: 'test',
    },
  });

  component.find('form').simulate('submit', { preventDefault () {} });

  expect(onSubmit.mock.calls[0][0]).toEqual({
    id: '1',
    workoutDate: '2017-08-08',
    remark: 'test',
    performances: [
      {
        event: events[0],
        set1: 1,
        set2: 2,
        set3: 3,
        set4: 4,
        set5: 5,
        value: 20,
      },
    ],
  });
});

it('simulate move list click', () => {
  const onMoveList = jest.fn();

  const component = mount(
    <Test>
      <PostForm
        events={events}
        initialValues={initialValues}
        formValues={initialValues}
        error={null}
        loading={false}
        onSubmit={() => {}}
        onDelete={() => {}}
        onMoveList={onMoveList}
      />
    </Test>
  );

  component.find('input[type="button"][value="List"]').at(0).simulate('click', {
    preventDefault: () => {},
  });

  expect(onMoveList).toBeCalled();
});

it('simulate delete click', () => {
  const onDelete = jest.fn();

  const component = mount(
    <Test>
      <PostForm
        events={events}
        initialValues={initialValues}
        formValues={initialValues}
        error={null}
        loading={false}
        onSubmit={() => {}}
        onDelete={onDelete}
        onMoveList={() => {}}
      />
    </Test>
  );

  component.find('input[type="button"][value="Delete"]').at(0).simulate('click', { preventDefault () {} });

  expect(onDelete).toBeCalled();
});
