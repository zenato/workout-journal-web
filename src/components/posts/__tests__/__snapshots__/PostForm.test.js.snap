// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without errors. 1`] = `
<Form
  onSubmit={[Function]}
>
  <FormGroup
    label="Workout Date"
  >
    <Field
      component={[Function]}
      name="workoutDate"
      type="date"
    />
  </FormGroup>
  <FieldArray
    component={[Function]}
    events={
      Array [
        Object {
          "id": "1",
          "name": "event-1",
        },
      ]
    }
    name="performances"
  />
  <FormGroup
    label="Remark"
  >
    <Field
      component={[Function]}
      name="remark"
      type="text"
    />
  </FormGroup>
  <styled.div>
    <Button
      color="primary"
      type="submit"
      value="Save"
    />
    <Button
      onClick={[Function]}
      value="List"
    />
  </styled.div>
</Form>
`;
